src/core/device.c

```diff
@@ -189,41 +182,64 @@ static int device_coldplug(Unit *u) {
-         *   will (re-)process all devices, and the Device.found and Device.state will be adjusted.
+         *   Unlike the other starting mode, Manager.honor_device_enumeration == false (maybe, it is better
+         *   to rename the flag) when device_coldplug() and device_catchup() are called. Hence, let's
+         *   conditionalize the operations by using the flag. After switch-root, systemd-udevd will
+         *   (re-)process all devices, and the Device.found and Device.state will be adjusted.
          *
          * - On reload or reexecute, we can trust enumerated_found, deserialized_found, and deserialized_state.
          *   Of course, deserialized parameters may be outdated, but the unit state can be adjusted later by
          *   device_catchup() or uevents. */
 
-        if (!m->honor_device_enumeration && !MANAGER_IS_USER(m)) {
-                found &= ~DEVICE_FOUND_UDEV; /* ignore DEVICE_FOUND_UDEV bit */
-                if (state == DEVICE_PLUGGED)
-                        state = DEVICE_TENTATIVE; /* downgrade state */
-                if (found == DEVICE_NOT_FOUND)
-                        state = DEVICE_DEAD; /* If nobody sees the device, downgrade more */
-        }
+        /* First, let's put the deserialized state and found mask into effect, if we have it. */
+        if (d->deserialized_state < 0)
+                return 0;
 
-        if (d->found == found && d->state == state)
+        if (d->found == d->deserialized_found && d->state == d->deserialized_state)
                 return 0;
 
-        d->found = found;
-        device_set_state(d, state);
+        d->found = d->deserialized_found;
+        device_set_state(d, d->deserialized_state);
         return 0;
 }
 
+static bool device_found_mask_on_catchup(Device *d) {
+        assert(d);
+
+        /* On switching root, the database for the device may be cleared, or the device itself may be
+         * unplugged. The function device_enumerate() cannot distinguish the two cases. Let's explicitly
+         * check if the device is still around. */
+
+        Manager *m = UNIT(d)->manager;
+        if (m->honor_device_enumeration || MANAGER_IS_USER(m))
+                /* Not switching root. Let's use the all bits. */
+                return DEVICE_FOUND_MASK;
+
+        if (FLAGS_SET(d->enumerated_mask, DEVICE_FOUND_UDEV))
+                /* The device is already enumerated. Let's use the all bits. */
+                return DEVICE_FOUND_MASK;
+
+        if (!d->path)
+                /* We have no way to check if the device exists. Assume the device is unplugged. */
+                return DEVICE_FOUND_MASK;
+
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        if (sd_device_new_from_path(&dev, d->path) < 0)
+                /* The device does not exist any more. */
+                return DEVICE_FOUND_MASK;
+
+        /* The device still exists, and we will receive uevent for the device when the device is re-processed
```

Comment:
IIUC we'd get here if `FLAGS_SET(d->enumerated_found, DEVICE_FOUND_UDEV)` is false (i.e. the device has been "forgotten" while switching root) but still exists in sysfs. [^1]

In this case you'll call `device_update_found_one()` with `DEVICE_FOUND_UDEV` cleared in the mask. In `device_found_changed()`, you'll end up setting the device state to `DEVICE_DEAD` because the previous `found` value has `DEVICE_FOUND_UDEV` set (see me remark for `device_coldplug()`), and the new `found` value has not.

Unless I'm completely on the wrong track, this is not what we want.

I like the basic idea to check device existence in sysfs before udev's coldplug processing occurs. But IMO we need to apply a different logic (roughly speaking, we should check sysfs if `serialized_found` indicates the devices is present but `enumerated_found` does not — I haven't thought it through yet).

[^1]: the device must have existed before switching root, otherwise we'd never know about it. While it can have been removed on the kernel side, udev can hardly have processed a REMOVE event yet at this point in time, when we're switching root.

Reply:
> IIUC we'd get here if `FLAGS_SET(d->enumerated_found, DEVICE_FOUND_UDEV)` is false (i.e. the device has been "forgotten" while switching root) but still exists in sysfs.

Right.

> In this case you'll call `device_update_found_one()` with `DEVICE_FOUND_UDEV` cleared in the mask. In `device_found_changed()`, you'll end up setting the device state to `DEVICE_DEAD` because the previous `found` value has `DEVICE_FOUND_UDEV` set (see me remark for `device_coldplug()`), and the new `found` value has not.
>
> Unless I'm completely on the wrong track, this is not what we want.

No, new `found` still has the DEVICE_FOUND_UDEV flag, as the mask does not have the bit.

Reply:
Right, sorry. I misread the code.

This means in the case described, you wouldn't clear the `DEVICE_FOUND_UDEV` bit. Thus in the case at hand, device state would only change if `DEVICE_FOUND_MOUNT` or `DEVICE_FOUND_SWAP` changed. Which basically can't happen because we wouldn't have (de)activated any mounts or swaps at the point in time when `enumerated_found` is initialized. IOW, the device state remains the same that had been set in `device_coldplug()`.

Reply:
> This means in the case described, you wouldn't clear the `DEVICE_FOUND_UDEV` bit. Thus in the case at hand, device state would only change if `DEVICE_FOUND_MOUNT` or `DEVICE_FOUND_SWAP` changed. Which basically can't happen because we wouldn't have (de)activated any mounts or swaps at the point in time when `enumerated_found` is initialized.

Right.

>  IOW, the device state remains the same that had been set in `device_coldplug()`.

No. `device_dispatch_io()` will drop the `DEVICE_FOUND_UDEV` flag on remove uevent or `SYSTEMD_READY=0` is set on add/change uevent.

Reply:
Unfortunately no. As mentioned in the above, I have no idea about fixing #23208 now...

Reply:
> (Sorry for bothering you here. I think it's important we clarify everything).

No problem. I am enjoying this whole discussion.